mmcv/ops/csrc/onnxruntime/cpu/modulated_deform_conv.cpp

```diff
@@ -0,0 +1,311 @@
+                for (int64_t k = 0; k < K; ++k) {
+                    y += A[k * M + m] * B[k * N + n];
+                }
+            }
+            if (trans_A && trans_B) { // KM, NK; TT
+                for (int64_t k = 0; k < K; ++k) {
+                    y += A[k * M + m] * B[n * K + k];
+                }
+            }
+            if (!trans_A && trans_B) { // MK, NK; NT
+                for (int64_t k = 0; k < K; ++k) {
+                    y += A[m * K + k] * B[n * K + k];
+                }
+            }
+            y *= alpha;
+            if (V) y += beta * V[n];
+            if (H) y += beta * H[m * N + n];
+            Y[m * N + n] = y;
+        }
+    }
+}
+
+void deformable_conv2d_ref_fp32(
+    const float *src,
+    const float *offset,
+    const float *mask,
+    const float *filter,
+    const float *bias,
+    const int64_t batch,
+    const int64_t src_c,
+    const int64_t src_h,
+    const int64_t src_w,
+    const int64_t dst_c,
+    const int64_t dst_h,
+    const int64_t dst_w,
+    const int64_t group,
+    const int64_t offset_group,
+    const int64_t channels,
+    const int64_t num_output,
+    const int64_t kernel_h,
+    const int64_t kernel_w,
+    const int64_t stride_h,
+    const int64_t stride_w,
+    const int64_t pad_h,
+    const int64_t pad_w,
+    const int64_t dilation_h,
+    const int64_t dilation_w,
+    float *dst)
+{
+    const int64_t ic_per_gp = channels / group;
+    const int64_t oc_per_gp = num_output / group;
+    float *columns = new float[ic_per_gp * kernel_h * kernel_w * dst_h * dst_w];
+    for (int64_t b = 0; b < batch; ++b) {
+        for (int64_t g = 0; g < group; ++g) {
+            deformable_im2col_2d(
+                src + b * src_c * src_h * src_w + g * ic_per_gp * src_h * src_w,
+                offset + b * offset_group * 2 * kernel_h * kernel_w * dst_h * dst_w,
+                mask + b * offset_group * kernel_h * kernel_w * dst_h * dst_w,
+                src_h, src_w,
+                kernel_h, kernel_w,
+                pad_h, pad_w,
+                stride_h, stride_w,
+                dilation_h, dilation_w,
+                ic_per_gp,
+                offset_group,
+                dst_h, dst_w,
+                mask != nullptr,
+                columns);
+            float *dst_ptr = dst + b * dst_c * dst_h * dst_w + g * oc_per_gp * dst_h * dst_w;
+            if (bias != nullptr) {
+                const float *bias_ptr = bias + g * oc_per_gp;
+                for (int64_t oc = 0; oc < oc_per_gp; ++oc) {
+                    for (int64_t hw = 0; hw < dst_h * dst_w; ++hw) {
+                        dst_ptr[oc * dst_h * dst_w + hw] = bias_ptr[oc];
+                    }
+                }
+            } else {
+                for (int64_t oc = 0; oc < oc_per_gp; ++oc) {
+                    for (int64_t hw = 0; hw < dst_h * dst_w; ++hw) {
+                        dst_ptr[oc * dst_h * dst_w + hw] = 0.0f;
+                    }
+                }
+            }
```

Comment:
Can we initialize `dst` before batch loop?
We can just init once and do `memcpy` for rest of dst.

Reply:
The `bias_ptr` depends on `g`. So we cannot initialize it just once and `memcpy` for rest of `dst_ptr`

Reply:
Ok, what about memset(0) when bias is nullptr?